File: src/main/scala/committee/nova/skillful/event/handler/ForgeEventHandler.scala
package committee.nova.skillful.event.handler

import committee.nova.skillful.Constants
import committee.nova.skillful.command.init.CommandInit
import committee.nova.skillful.event.handler.CapabilityHandler.skillfulCap
import committee.nova.skillful.event.impl.{SkillLevelEvent, SkillXpEvent}
import committee.nova.skillful.implicits.Implicits.PlayerEntityImplicit
import committee.nova.skillful.player.capabilities.impl.Skills
import committee.nova.skillful.storage.SkillfulStorage
import net.minecraft.entity.Entity
import net.minecraft.entity.player.{PlayerEntity, ServerPlayerEntity}
import net.minecraft.network.play.server.SPlaySoundEffectPacket
import net.minecraft.util.text.{TextFormatting, TranslationTextComponent}
import net.minecraft.util.{ResourceLocation, SoundCategory, SoundEvents, Util}
import net.minecraftforge.common.util.FakePlayer
import net.minecraftforge.event.TickEvent.Phase
import net.minecraftforge.event.entity.living.LivingEntityUseItemEvent
import net.minecraftforge.event.entity.player.PlayerEvent.{Clone, PlayerChangedDimensionEvent, PlayerLoggedInEvent, PlayerRespawnEvent}
import net.minecraftforge.event.entity.player.PlayerWakeUpEvent
import net.minecraftforge.event.{AttachCapabilitiesEvent, RegisterCommandsEvent, TickEvent}
import net.minecraftforge.eventbus.api.SubscribeEvent

import scala.collection.mutable

class ForgeEventHandler {
  @SubscribeEvent
  def onAttachCaps(event: AttachCapabilitiesEvent[Entity]): Unit = {
    event.getObject match {
      case _: PlayerEntity => event.addCapability(new ResourceLocation(Constants.MODID, Constants.MODID), new Skills.Provider)
      case _ =>
    }
  }

  @SubscribeEvent
  def onRegisterCommand(e: RegisterCommandsEvent): Unit = CommandInit.init(e.getDispatcher)

  @SubscribeEvent
  def onPlayerTick(event: TickEvent.PlayerTickEvent): Unit = {
    if (event.phase == Phase.START) return
    event.player match {
      case _: FakePlayer =>
      case p: ServerPlayerEntity => p.getSkills.getSkillInfos.foreach(t => if (t.isActive && t.tick()) t.removePlayer(p))
      case _ =>
    }
  }

  @SubscribeEvent
  def onPlayerLogin(event: PlayerLoggedInEvent): Unit = {
    event.getPlayer match {
      case _: FakePlayer =>
      case p: ServerPlayerEntity =>
        p.applySkillAttrs()
        p.syncSkills()
      case _ =>
    }
  }

  @SubscribeEvent
  def onPlayerChangedDimension(event: PlayerChangedDimensionEvent): Unit = {
    if (event.getPlayer.isFake) return
    event.getPlayer.clearSkillInfoCache()
    event.getPlayer.syncSkills()
    event.getPlayer.applySkillAttrs()
  }

  @SubscribeEvent
  def onPlayerRespawn(event: PlayerRespawnEvent): Unit = {
    if (event.getPlayer.isFake) return
    event.getPlayer.clearSkillInfoCache()
    event.getPlayer.syncSkills()
    event.getPlayer.applySkillAttrs()
  }

  @SubscribeEvent
  def onClone(event: Clone): Unit = {
    val oldPlayer = event.getOriginal
    if (oldPlayer.isFake) return
    val newPlayer = event.getPlayer
    val cap = skillfulCap
    val storage = skillfulCap.getStorage
    oldPlayer.getCapability(cap).ifPresent(o => newPlayer.getCapability(cap).ifPresent(n =>
      storage.readNBT(cap, n, null, storage.writeNBT(cap, o, null))
    ))
  }

  @SubscribeEvent
  def onXpChanged(event: SkillXpEvent.Post): Unit = if (!event.getPlayer.isFake) event.getPlayer.sendSkillInfo(event.getSkillInstance, event.getAmount)

  @SubscribeEvent
  def onLevelChanged(event: SkillLevelEvent): Unit = {
    val player = event.getPlayer
    if (player.isFake) return
    val isUp = event.isUp
    event.getSkillInstance.getSkill match {
      case c if c.shouldActOnLevelChange => c.actOnLevelChange(player, event.getSkillInstance, isUp)
      case _ =>
    }
    if (isUp && event.getSkillInstance.getCurrentLevel != 1)
      player.connection.sendPacket(new SPlaySoundEffectPacket(SoundEvents.ENTITY_PLAYER_LEVELUP, SoundCategory.PLAYERS,
        player.getPosX, player.getPosY, player.getPosZ, 1.0F, 1.0F))
  }

  @SubscribeEvent
  def onWakeUp(event: PlayerWakeUpEvent): Unit = {
    if (!(!event.updateWorld() && !event.wakeImmediately())) return
    val player = event.getPlayer
    if (player.isFake) return
    val buffer = new mutable.ArrayBuffer[(ResourceLocation, Int)]
    player match {
      case p: ServerPlayerEntity =>
        p.getSkills.getSkills.foreach(i => {
          i.getSkill match {
            case x if x.shouldChangeXPAfterSleep =>
              val v = x.changeXPAfterSleep(p, i)
              if (v != 0) {
                i.addXp(p, v)
                buffer.+=((x.getId, v))
              }
            case _ =>
          }
        })
        if (buffer.nonEmpty) {
          p.sendMessage(new TranslationTextComponent("info.skillful.wakeup.summary").mergeStyle(TextFormatting.LIGHT_PURPLE), Util.DUMMY_UUID)
          buffer.foreach(x => {
            val c = x._2
            p.sendMessage(new TranslationTextComponent(s"info.skillful.wakeup.${if (c > 0) "increase" else "decrease"}",
              new TranslationTextComponent(s"skill.${x._1.getNamespace}.${x._1.getPath}"),
              c.abs.toString
            ).mergeStyle(if (c > 0) TextFormatting.GREEN else TextFormatting.RED), Util.DUMMY_UUID)
          })
        }
      case _ =>
    }
  }

  @SubscribeEvent
  def onFoodEaten(e: LivingEntityUseItemEvent.Finish): Unit = {
    if (!e.getEntityLiving.isInstanceOf[ServerPlayerEntity] || e.getEntityLiving.isInstanceOf[FakePlayer]) return
    val stack = e.getItem
    SkillfulStorage.applyFoodEffect(e.getEntityLiving.asInstanceOf[ServerPlayerEntity], stack)
  }
}

File: src/main/java/committee/nova/skillful/common/event/handler/ForgeEventHandler.java
package committee.nova.skillful.common.event.handler;

import committee.nova.skillful.Skillful;
import committee.nova.skillful.common.cap.skill.ISkills;
import committee.nova.skillful.common.cap.skill.Skills;
import committee.nova.skillful.common.command.CommandManager;
import committee.nova.skillful.common.util.Utilities;
import net.minecraft.ChatFormatting;
import net.minecraft.network.chat.TranslatableComponent;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraftforge.common.capabilities.RegisterCapabilitiesEvent;
import net.minecraftforge.common.util.FakePlayer;
import net.minecraftforge.common.util.LazyOptional;
import net.minecraftforge.event.AttachCapabilitiesEvent;
import net.minecraftforge.event.RegisterCommandsEvent;
import net.minecraftforge.event.entity.player.PlayerEvent;
import net.minecraftforge.event.entity.player.PlayerWakeUpEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;

@Mod.EventBusSubscriber
public class ForgeEventHandler {
    @SubscribeEvent
    public static void onRegisterCap(RegisterCapabilitiesEvent event) {
        event.register(Skills.Impl.class);
    }

    @SubscribeEvent
    public static void onAttachCap(AttachCapabilitiesEvent<Entity> event) {
        if (!(event.getObject() instanceof Player player)) return;
        if (player instanceof FakePlayer) return;
        event.addCapability(new ResourceLocation(Skillful.MODID, "skills"), new Skills.Provider());
    }

    @SubscribeEvent
    public static void onClone(PlayerEvent.Clone event) {
        final Player original = event.getOriginal();
        original.reviveCaps();
        final LazyOptional<ISkills> oldCap = original.getCapability(Skills.SKILLS_CAPABILITY);
        final LazyOptional<ISkills> newCap = event.getEntity().getCapability(Skills.SKILLS_CAPABILITY);
        newCap.ifPresent((n) -> oldCap.ifPresent((o) -> n.deserializeNBT(o.serializeNBT())));
        original.invalidateCaps();
    }

    @SubscribeEvent
    public static void onLogin(PlayerEvent.PlayerLoggedInEvent event) {
        checkAndSync(event.getPlayer());
    }

    @SubscribeEvent
    public static void onChangedDimension(PlayerEvent.PlayerChangedDimensionEvent event) {
        checkAndSync(event.getPlayer());
    }

    @SubscribeEvent
    public static void onRespawn(PlayerEvent.PlayerRespawnEvent event) {
        checkAndSync(event.getPlayer());
    }

    @SubscribeEvent
    public static void onWakeup(PlayerWakeUpEvent event) {
        if (event.updateWorld() || event.wakeImmediately()) return;
        if (!(event.getEntity() instanceof ServerPlayer p) || p instanceof FakePlayer) return;
        p.displayClientMessage(new TranslatableComponent("msg.skillful.sleep")
                .withStyle(ChatFormatting.AQUA), false);
        p.getCapability(Skills.SKILLS_CAPABILITY)
                .ifPresent(i -> i.getSkills().forEach(s -> s.getSkill().onWakeup(p, s)));
    }

    @SubscribeEvent
    public static void onRegisterCmd(RegisterCommandsEvent event) {
        CommandManager.init(event.getDispatcher());
    }

    private static void checkAndSync(Player player) {
        if (!(player instanceof ServerPlayer s) || s instanceof FakePlayer) return;
        Utilities.checkSkills(s);
        Utilities.syncSkills(s);
    }
}
